### PR TITLE
Add TreeLSTM links

### PR DESCRIPTION
**Pull request by soskek**
I implemented two variants of TreeLSTM (http://www.aclweb.org/anthology/P15-1150), Child-Sum TreeLSTM and N-ary TreeLSTM, from discussions of https://github.com/pfnet/chainer/issues/2592.
Note that these are written with basic functions and links directly, so are not well-optimized for efficiency.

### CODE REVIEW

**Comment by soskek** [comment]
I finished the implementation and unit tests.
Next, I will examine this on the chainer's sentiment example using tree-based recursive structure.

**Comment by soskek** [comment]
I confirmed that TreeLSTM outperformed a baseline tanh-based recursive neural net. The baseline with 30 units achieved 43.42% accuracy of root nodes at dev set, N-ary TreeLSTM with 30 units achieved 45.87%. It works well.
As a further analysis, the baseline with 60 or 120 got overfitting and decreased performance at dev set. However, N-ary TreeLSTM with 120 units achieved 46.78% accuracy.
(In the original paper of TreeLSTM, dropout and different learning hyperparameters are used. But, in these, I did not modify the current sentiment example of Chainer.)
I also confirmed the TreeLSTM is much slower than the baseline, at least 5 times, in a cpu experiment without batch computation.

In the experiments, I used the Chain with TreeLSTM below instead of current `RecursiveNet`, in sentiment example,
```
class RecursiveTreeLSTMNet(chainer.Chain):

    def __init__(self, n_vocab, n_units):
        super(RecursiveTreeLSTMNet, self).__init__(
            embed=L.EmbedID(n_vocab, n_units),
            treelstm=L.NaryTreeLSTM(n_units, n_units, n_ary=2),
            w=L.Linear(n_units, n_label))

    def leaf(self, x):
        emb = self.embed(x)
        zero = self.xp.zeros(emb.shape, 'f')
        c, h = self.treelstm(zero, zero, zero, zero, emb)
        return (c, h)

    def node(self, left, right):
        c_left, h_left = left
        c_right, h_right = right
        zero = self.xp.zeros(c_left.shape, 'f')
        c, h = self.treelstm(c_left, c_right, h_left, h_right, zero)
        return (c, h)

    def label(self, v):
        h = v[1]
        return self.w(h)
```

**Comment by unnonouno** [comment]
Great work!!

**Comment by jekbradbury** [comment]
This may be useful to you guys -- I think it's a slightly different structure but it's worked pretty well for me (and it has a fused kernel). The docstrings are mostly copied from the old lstm docstrings so some things might be wrong.
https://gist.github.com/jekbradbury/a80b430564289fe9dd0cc6ac978ab052

**Comment by soskek** [comment]
Thank you for the useful reference code!! I am considering to transfer/merge into the current implementation.
Before that, could you discuss good specifications of Chainer's TreeLSTMs with me?

In my understanding, the differences of 2-ary TreeLSTM among (i) Tai's, (ii) Bowman's, and (iii) yours are as follows:
- In (i), the matrices of two forget gates, for left-child and right-child, are distinct for h of two child-nodes. So, the total number of matrices are 14; `4 for x-to-a/i/o/**f_{both}**, 5 for h_{left}-to-a/i/o/f_{left}/f_{right}, 5 for h_{left}-to-a/i/o/f_{left}/f_{right}`.
- In (ii), additionally, x also makes a distinction between left and right. So, the total number of matrices are 15; `5 for x-to-a/i/o/**f_{left}/f_{right}**, 5 for h_{left}-to-a/i/o/f_{left}/f_{right}, 5 for h_{left}-to-a/i/o/f_{left}/f_{right}`. And, your code (iii) also obey this, as you commented "f1 and f2 can depend in different ways on the LSTM input." in the docstring.
- ((ii) paper has one misprint at `h = o * c`. `h = o * tanh(c)` is correct. This is confirmed at https://github.com/sleepinyourhat/spinn/blob/master/python/spinn/fat_stack.py.)
- You commented "10/9 as many independent parameters as Tai" in (iii)'s docstring. The correct one is not "10/9" but "15/14", I suppose. Would you please re-confirm my understanding?

(Here, I use `a` consistently instead of `u` in Tai's paper and `g` in Bowman's.)

In summary,
which do we implement, Tai's or Bowman's?
Bowman's one is little more powerful and simple, and I think people may not hesitate to use Bowman's one.
If you agree with me, I will implement Bowman's N-ary TreeLSTM.

**Comment by jekbradbury** [comment]
Yes, I think all of that is accurate. (ii)/(iii) is significantly easier to implement than (i) as a combination of linear layers + an elementwise CUDA kernel and I would be very surprised if they performed differently.
And I was also tripped up by the typo in Bowman's paper, which has been fixed in some versions of the PDF but not in others. I clarified it with Sam last year.

**Comment by soskek** [comment]
Thank you for the early response. Oh, I found it was fixed on the latest version on arxiv https://arxiv.org/pdf/1603.06021.pdf, as you say.
OK, everything got clarified, and I begin to introduce Bowman's version by cuda kearnel for N-ary TreeLSTM.

**Comment by soskek** [comment]
First, I implemented Bowman's Binary TreeLSTM with modifications of the reference code.
I am continuing to re-extend it to variable N-ary TreeLSTM. After that, I will also modify Child-Sum TreeLSTM with cuda kernel.

File: chainer/functions/activation/tree_lstm.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+                    {}
+                '''.format(gf_calc_str, gc_calc_str),
+                'treelstm_bwd', preamble=preamble)(
+                    self.c, gc, gh, a, i, o,
+                    *(list(cs) + fs + [ga, gi, go] + gcs + gfs))
+
+        return list(gcs) + [gx]
+
+
+def tree_lstm(*inputs):
```
Comment (unnonouno) [inline chainer/functions/activation/tree_lstm.py]: It is a little bit difficult to understand its specification.  How about like this? 
```def tree_lstm(cs, x):```

File: chainer/links/connection/tree_lstm.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        super(ChildSumTreeLSTM, self).__init__(
+            W_x=linear.Linear(in_size, 4 * out_size),
+            W_h_aio=linear.Linear(out_size, 3 * out_size, nobias=True),
+            W_h_f=linear.Linear(out_size, out_size, nobias=True),
+        )
+        self.in_size = in_size
+        self.state_size = out_size
+        utils.experimental('chainer.links.tree_lstm.py')
+
+    def __call__(self, *cshsx):
```
Comment (unnonouno) [inline chainer/links/connection/tree_lstm.py]: How about getting lists directly?
```def __call__(self, cs, hs, x)```

File: chainer/links/connection/tree_lstm.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+
+        """
+
+        cs = cshsx[:len(cshsx) // 2]
+        hs = cshsx[len(cshsx) // 2:-1]
+        x = cshsx[-1]
+
+        if x is None:
+            x = self.xp.zeros(
+                (cs[0].shape[0], self.in_size), dtype=cs[0].dtype)
```
Comment (unnonouno) [inline chainer/links/connection/tree_lstm.py]: Check if `len(cs) > 0`

File: chainer/links/connection/tree_lstm.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+        W_x_in = self.W_x(x)
+        W_x_aio_in, W_x_f_in = split_axis.split_axis(
+            W_x_in, [3 * self.state_size], axis=1)
+
+        if any(h is None for h in hs):
```
Comment (unnonouno) [inline chainer/links/connection/tree_lstm.py]: Is it possible to extract a method for hs and cs?

File: chainer/functions/activation/tree_lstm.py (hunk 1 above)
Comment (soskek) [inline chainer/functions/activation/tree_lstm.py]: N-ary TreeLSTM is a purely generalized extension of LSTM, because 1-ary TreeLSTM equals to LSTM. (In fact, the current implementation does not allow N = 1 with an assertion.)
Thus, I defined the specification to make a consistency of input arguments with `chainer.functions.lstm(c_prev, x)` (and old `chainer.functions.slstm(c_prev1, c_prev2, x1, x2)`).
However, what you say is also reasonable. I can either modify or keep this. What do you think again for this?

File: chainer/links/connection/tree_lstm.py (hunk 2 above)
Comment (soskek) [inline chainer/links/connection/tree_lstm.py]: This also follows the reason above for the functions and `chainer.links.StatelessLSTM.__call__(c, h, x)`.

File: chainer/links/connection/tree_lstm.py (hunk 3 above)
Comment (soskek) [inline chainer/links/connection/tree_lstm.py]: OK, and I will add a test for None input soon.

File: chainer/links/connection/tree_lstm.py (hunk 4 above)
Comment (soskek) [inline chainer/links/connection/tree_lstm.py]: Sure. I got it.

File: chainer/functions/activation/tree_lstm.py (hunk 1 above)
Comment (soskek) [inline chainer/functions/activation/tree_lstm.py]: I discussed this issue w/ @unnonouno offline. We decided to keep the consistency. So this specification will be unchanged.

File: chainer/functions/activation/tree_lstm.py [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+from chainer import cuda
+from chainer import function
+from chainer.utils import type_check
+
+
+def _extract_gates(x, n_split=5):
+    """Extract gates by split.
+
+    This is different from ``_extract_gates`` in lstm.py,
+    which is as follows
```
Comment (unnonouno) [inline chainer/functions/activation/tree_lstm.py]: like this
```
which is as follows::

        r = x.reshape((x.shape[0], x.shape[1] // 4, 4) + x.shape[2:])
        return (r[:, :, i] for i in six.moves.range(4))

In other ...
```

File: chainer/functions/activation/tree_lstm.py [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+        r = x.reshape((x.shape[0], x.shape[1] // 4, 4) + x.shape[2:])
+        return (r[:, :, i] for i in six.moves.range(4))
+    ::
+    In other words, it thinly slices x and merge them,
+    while this thickly slices x.
+
+    """
+    r = x.reshape(
+        (x.shape[0], n_split, x.shape[1] // n_split) + x.shape[2:])
+    return (r[:, i, :] for i in six.moves.range(n_split))
```
Comment (unnonouno) [inline chainer/functions/activation/tree_lstm.py]: How about using `split_axis`?

File: chainer/functions/activation/tree_lstm.py [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+
+        return list(gcs) + [gx]
+
+
+def tree_lstm(*inputs):
+    """TreeLSTM unit as an activation function.
+
+    This function implements TreeLSTM units both for
+    N-ary TreeLSTM and Child-Sum TreeLSTM.
+    Let the children cell states :math:`c_{\\text{1}}` `:math:c_{\\text{2}}`
```
Comment (unnonouno) [inline chainer/functions/activation/tree_lstm.py]: How abou using one LaTeX <pre>:math:\`c_{\\text{1}}, c_{\\text{2}}, \\dots, c_{\\text{N}}\`</pre>

File: chainer/functions/activation/tree_lstm.py (hunk 6 above)
Comment (soskek) [inline chainer/functions/activation/tree_lstm.py]: I am just following `lstm`'s implementation (https://github.com/pfnet/chainer/blob/_v2/chainer/functions/activation/lstm.py). So I can change it.
But, if using `split_axis`, squeezing axis=1 seems needed after split.
For example,
```
(F.squeeze(sp, axis=1) for sp in F.split_axis(r, n_split, axis=1))
```

I am not sure that which is better for efficiency or readability.

**Comment by soskek** [comment]
All comments are fixed.
I think appveyor's errors look not due to this PR.

**Comment by unnonouno** [comment]
Thank you! LGTM. Please send it to `origin/master`.

File: chainer/functions/activation/tree_lstm.py [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+    .. admonition:: Example
+
+        Assuming ``y`` is the current input signal, ``c`` is the previous cell
+        state, and ``h`` is the previous output signal from an
+        ``n_ary_tree_lstm`` function.
+        Each of ``y``, ``c`` and ``h`` has ``n_units`` channels.
+        Using 2-ary (binary) TreeLSTM,
+        most typical preparation of ``x`` is:
+
+        >>> model = chainer.Chain(w=F.Linear(n_units, 5 * n_units),
```
Comment (unnonouno) [inline chainer/functions/activation/tree_lstm.py]: Please make it runnable. Run `make test` under `docs` directory. I mean, `n_units`, `y`, `h1` and `h2` are undefined here.

**Comment by unnonouno** [comment]
Great job!!
